tests/device_tests/cardano/test_address_public_key.py

```diff
@@ -0,0 +1,139 @@
+# This file is part of the Trezor project.
+#
+# Copyright (C) 2012-2019 SatoshiLabs and contributors
+#
+# This library is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License version 3
+# as published by the Free Software Foundation.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the License along with this library.
+# If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
+
+import pytest
+
+from trezorlib.cardano import (
+    NETWORK_IDS,
+    create_address_parameters,
+    get_address,
+    get_public_key,
+)
+from trezorlib.messages import CardanoAddressType
+from trezorlib.tools import parse_path
+
+from ...common import parametrize_using_common_fixtures
+
+pytestmark = [
+    pytest.mark.altcoin,
+    pytest.mark.cardano,
+    pytest.mark.skip_t1,
+    pytest.mark.skip_ui,
+]
+
+
+@parametrize_using_common_fixtures(
+    "cardano/get_address.json", "cardano/get_address.slip39.json"
+)
+def test_cardano_get_address(client, parameters, result):
+    address = get_address(
+        client,
+        address_parameters=create_address_parameters(
+            address_type=CardanoAddressType.BYRON,
+            address_n=parse_path(parameters["path"]),
+        ),
+        protocol_magic=parameters["protocol_magic"],
+        network_id=NETWORK_IDS["mainnet"],
```

Comment:
shouldn't this also be part of the jsons?

Reply:
actually ISTM this test should be called `test_cardano_get_address_byron`, and the JSONs renamed appropriately

Reply:
on second thought, maybe ALL the individual (Byron, Base, Enterprise, Pointer...) subvectors should be one single test `test_cardano_get_address`, with both `address_type` and `network_id` provided by the json.

there will still be separate files for `get_address_byron.json` and `get_address_byron_slip39.json` because they are using different mnemonics (right?), but otherwise there could easily be one vector file -- or multiple in the same format applicable for the same testcase

Reply:
> shouldn't this also be part of the jsons?

Yes. Will fix.

> on second thought, maybe ALL the individual (Byron, Base, Enterprise, Pointer...) subvectors should be one single test test_cardano_get_address, with both address_type and network_id provided by the json.

Now that I look into it, it makes sense. But I would still stick with numerous json files. Is that what you meant? Having X json files but one test case.

Reply:
> Having X json files but one test case.

yes, as appropriate. good thing the decorator takes multiple arguments :)

(do we want to introduce wildcards for this?)

Reply:
As I am looking into this the problem is that different Cardano address types require different parameters.

For example `CardanoAddressType.BASE` requires `staking_key_hash` and `POINTER` block_index etc. I guess we could solve this by some magic (since the parameter names are the same in `create_address_parameters` as they are in fixtures but maybe it is too much of a hassle?

Reply:
i would do `parameters["address_parameters"].get("whichever_value")` for _all_ the non-mandatory arguments?

Reply:
Is there a shorter way to write this:

```
staking_path = parameters.get("staking_path")
    if staking_path is not None:
        staking_path = parse_path(staking_path)
```
?

I am sure there is but I am bad at it :laughing:.

Reply:
Other than that please check out e89fbe1d9ef3c5ab9b48ae2359481f3cdc928d12.

Reply:
> (do we want to introduce wildcards for this?)

Maybe? Or when we find it necessary? Because I kind of like how explicit it is. https://github.com/trezor/trezor-firmware/commit/e89fbe1d9ef3c5ab9b48ae2359481f3cdc928d12#diff-18e7411a22974b97bcf18c3825e30a5aR48-R54

Reply:
> Is there a shorter way

`parse_path(parameters.get("staking_path"), "")` happens to do what we want here, because `path=[]` is equivalent to `path=None` by the time the data hits protobuf

in general, probably not. you could do `parse_path(parameters["bla"]) if "bla" in parameters else None`, but that's not amazing either

Reply:
That's not gonna work for `staking_key_hash` though. `bytes.fromhex(parameters.get("staking_key_hash", ""))` passes `""` as an argument but I need `None` and `bytes.fromhex(None)` is not valid as well. So I guess I need to use the latter anyway..

Reply:
Done in bbf842d193e4660fef32b2afe12f6196164df878